File: codejam/2008/round_1c/text_messaging_outrage/solution.hs
import Data.List (intersperse, sort)

infixr 0 <|
(<|) f x = f x
(<||) f g x = f <| g x

optimalAssignmentCost :: Int -> [Int] -> Int
optimalAssignmentCost k frequencies = sum
    <| map (uncurry (*))
    <| zip (reverse <| sort frequencies)
    <| [1..] >>= replicate k

solve :: IO String
solve = do
    line <- getLine
    let [_, k, _] = map (read :: String -> Int) <| words line
    line <- getLine
    let frequencies = map (read :: String -> Int) <| words line
    return <| show <| optimalAssignmentCost k frequencies

main :: IO ()
main = (readLn :: IO Int)
    >>= sequence <|| flip replicate solve
    >>= putStrLn
        <|| concat
        <|| intersperse "\n"
        <|| map (uncurry (++))
        <|| zip (map (("Case #" ++) <|| (++ ": ") <|| show) [1..])

File: codejam/2008/round_1c/text_messaging_outrage/solution.rs
lib::run!();

struct TestCase {
    k: usize,
    frequencies: Vec<usize>,
}

fn read() -> TestCase {
    lib::input!([usize; 3] as [_, k, _]);
    lib::input!(Vec<usize> as frequencies);
    TestCase { k, frequencies }
}

fn solve(TestCase { k, mut frequencies }: TestCase) -> String {
    frequencies.sort();
    frequencies.reverse();
    format!(
        "{}",
        frequencies
            .into_iter()
            .enumerate()
            .map(|(i, f)| f * (1 + i / k))
            .sum::<usize>()
    )
}
